{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# class perform sequence alignment using dynamic programming.\n",
    "\n",
    "class DynamicProgrammer(object):\n",
    "    int_to_char = {0:'A', 1:'C', 2:'G', 3:'T', 4:'N'}\n",
    "    def __init__(self, seq1, seq2,indel,scoring):\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.indel = indel\n",
    "        self.scoring = scoring\n",
    "        self.D = None\n",
    "\n",
    "    def find_gobal_alignment(self):\n",
    "        self.D = np.zeros((self.seq1.size+1, self.seq2.size+1), dtype=np.int16)\n",
    "        self._compute_array()\n",
    "        # self._print_sequences()\n",
    "        return self._get_final_score()\n",
    "\n",
    "    def _compute_array(self):\n",
    "        for i in xrange(self.seq1.size+1):\n",
    "            self.D[i,0] = i*self.indel\n",
    "        for j in xrange(self.seq2.size+1):\n",
    "            self.D[0,j] = j*self.indel\n",
    "        for i in xrange(1, self.seq1.size+1):\n",
    "            for j in xrange(1, self.seq2.size+1):\n",
    "                self.D[i,j] = max(  self.D[i-1, j-1] + self._get_score(i, j),\n",
    "                                    self.D[i-1, j] + self.indel,\n",
    "                                    self.D[i, j-1] + self.indel)\n",
    "    def _get_score(self, i, j):\n",
    "        ''' The indexing is quite tricky because the matrix as one more row & column.\n",
    "        That causes a shift between the matrix index and the sequence indices.\n",
    "        Therefore, to obtain the correct nucleotide in the sequence, we must\n",
    "        substract 1 to the matrix index. '''\n",
    "        return self.scoring[self.seq1[i-1], self.seq2[j-1]]\n",
    "    \n",
    "    def _get_final_score(self):\n",
    "        i = self.seq1.size\n",
    "        j = self.seq2.size\n",
    "        return self.D[i,j]\n",
    "    \n",
    "    def _get_aligned_pair(self, i, j):\n",
    "        n1 = int_to_char[self.seq1[i-1]] if i>0 else '_'\n",
    "        n2 = int_to_char[self.seq2[j-1]] if j>0 else '_'\n",
    "        return (n1, n2)\n",
    "\n",
    "    def _traceback(self):\n",
    "        alignment= []\n",
    "        i = self.seq1.size\n",
    "        j = self.seq2.size\n",
    "        while i >0 and j>0:\n",
    "            if self.D[i-1, j-1] + self._get_score(i, j) == self.D[i,j]:\n",
    "                alignment.append(self._get_aligned_pair(i, j))\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif self.D[i-1, j] + self.indel == self.D[i,j]:\n",
    "                alignment.append(self._get_aligned_pair(i, 0))\n",
    "                i -= 1\n",
    "            else:\n",
    "                alignment.append(self._get_aligned_pair(0, j))\n",
    "                j -= 1\n",
    "        while i > 0:\n",
    "            alignment.append(self._get_aligned_pair(i, 0))\n",
    "            i -= 1\n",
    "        while j > 0:\n",
    "            alignment.append(self._get_aligned_pair(0, j))\n",
    "            j -= 1\n",
    "        alignment.reverse()\n",
    "        return alignment\n",
    "    \n",
    "    def _print_sequences(self):\n",
    "        pairs = self._traceback()\n",
    "        top_seq = []\n",
    "        bottom_seq = []\n",
    "        for (b, t) in pairs:\n",
    "            bottom_seq.append(b)\n",
    "            top_seq.append(t)\n",
    "        for n in top_seq:\n",
    "            print n,\n",
    "        print ' '\n",
    "        for n in bottom_seq:\n",
    "            print n,\n",
    "        print (\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above class is a class to perform dynamic programming for sequence alignment. Notice that the class have to used within barcode_compare function defined below. **barcode_compare** function is to compare two sequences using dynamic programming ,difflib or other methods. The scoring matrix and penalty score could be altered. By default, \n",
    "* DP (dynamic programming) is to final optimal global alignment of two sequences.\n",
    "    * indel = -2\n",
    "    * match = 1\n",
    "    * mimatch = -1\n",
    "    * N = 0\n",
    "* difflib is a python module to compare two strings. N won't be considered specifically. Only uses default score in **difflib** module.\n",
    "* other is a method to compare two sequences base by base from 5' to 3'. Notice two sequences should be in the same length. Otherwise error will be raised.\n",
    "    * match = 1\n",
    "    * mimatch = -1\n",
    "    * N = 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function perform barcode comaprison and return the similarity score.\n",
    "\n",
    "def barcode_compare (seq1, seq2, method = \"DP\"):\n",
    "    '''\n",
    "    Input: \n",
    "    seq1 is the a barcode sequence.\n",
    "    seq2 is another barcode sequence.\n",
    "    There are three methods to compare the similarity of two barcodes.\n",
    "     - DP: dynamic programming\n",
    "     - difflib: a python package to compare two string similarity\n",
    "     - other: compare two barcode base by base from 5' to 3'\n",
    "    Output:\n",
    "    final_score describe the differences.\n",
    "     - DP: similarity matrix based dynamic programming score.\n",
    "     - difflib. raito represent the similarity.\n",
    "     - other: score calculated base by base.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    char_to_int = {'A':0, 'C':1, 'G':2, 'T':3, 'N':4}\n",
    "    \n",
    "    if method == \"DP\":\n",
    "        # make sure the sequence comapred is DNA sequences. (N is permitted).\n",
    "        valid_dna = \"ATCGN\"\n",
    "        is_DNA1 = all(i in valid_dna for i in seq1)\n",
    "        is_DNA2 = all(i in valid_dna for i in seq1)\n",
    "        assert(is_DNA1 & is_DNA2)\n",
    "\n",
    "        \n",
    "        # define penalty score.\n",
    "        indel = -2 \n",
    "        scoring = array([[1,-1,-1,-1,0],\n",
    "                         [-1,1,-1,-1,0],\n",
    "                         [-1,-1,1,-1,0],\n",
    "                         [-1,-1,-1,1,0],\n",
    "                         [0,0,0,0,0]])\n",
    "\n",
    "        s1 = array([char_to_int[i] for i in seq1], dtype=np.int16)\n",
    "        s2 = array([char_to_int[i] for i in seq2], dtype=np.int16)\n",
    "        aligner = DynamicProgrammer(s1, s2,indel,scoring)\n",
    "        final_score = aligner.find_gobal_alignment()\n",
    "    elif method == \"difflib\":\n",
    "        s = df.SequenceMatcher(lambda x: x == \"N\",seq1,seq2)\n",
    "        final_score = s.ratio()\n",
    "    else:\n",
    "        assert(len(seq1) == len(seq2))\n",
    "        final_score = 0\n",
    "        match_score = 1\n",
    "        N_score = 0\n",
    "        mimatch_score = -1\n",
    "        for i in range(len(seq1)):\n",
    "            if seq1[i] != \"N\" and seq2[i] != \"N\":\n",
    "                if seq1[i] == seq2[i]:\n",
    "                    final_score += match_score\n",
    "                else:\n",
    "                    final_score += mimatch_score\n",
    "            else:\n",
    "                final_score += N_score\n",
    "    \n",
    "    return(final_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we will show some examples to observe how to works for different barcode comparison. From the result we could see that (for barcode with length eight):\n",
    "* perfect match : 8/1/8\n",
    "* perfect match w N:  N: 7/0.875/7\n",
    "* one mismatch: 6/0.875/6\n",
    "* one indel: 5/0.933/NA \n",
    "* one mismatch with N: 5/0.625/5\n",
    "* two mismatches: 4/0.625/4\n",
    "\n",
    "** Therefore, a single-end cut-off could be 6/0.875/6.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perfectly match r1:8\n",
      "Perfectly match r2:1.0\n",
      "Perfectly match r3:8\n",
      "Perfectly match but with 1N r1:7\n",
      "Perfectly match but with 1N r2:0.875\n",
      "Perfectly match but with 1N r3:7\n",
      "One mismatch r1:6\n",
      "One mismatch r2:0.875\n",
      "One mismatch r3:6\n",
      "One indel r1:5\n",
      "One indel r2:0.933333333333\n",
      "One mismatch with 1N r1:5\n",
      "One mismatch with 1N r2:0.625\n",
      "One mismatch with 1N r3:5\n",
      "Two mismatch r1:4\n",
      "Two mismatch r2:0.625\n",
      "Two mismatch r3:4\n"
     ]
    }
   ],
   "source": [
    "# an example show barcode_compare()\n",
    "import difflib as df\n",
    "import numpy as np\n",
    "from numpy import array\n",
    "import re\n",
    "    \n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGGACG\"\n",
    "r1 = barcode_compare(bcode1,bcode2) # dynamic comparison.\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\") # diff.ratio comparison.\n",
    "r3= barcode_compare(bcode1,bcode2,method=\"other\") # custermized one-to-one comparison.\n",
    "\n",
    "print(\"Perfectly match r1:\" + str(r1))\n",
    "print(\"Perfectly match r2:\" + str(r2))\n",
    "print(\"Perfectly match r3:\" + str(r3))\n",
    "\n",
    "\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGNACG\"\n",
    "r1 = barcode_compare(bcode1,bcode2)\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\")\n",
    "r3= barcode_compare(bcode1,bcode2,method=\"other\")\n",
    "\n",
    "print(\"Perfectly match but with 1N r1:\" + str(r1))\n",
    "print(\"Perfectly match but with 1N r2:\" + str(r2))\n",
    "print(\"Perfectly match but with 1N r3:\" + str(r3))\n",
    "\n",
    "\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGCACG\"\n",
    "r1 = barcode_compare(bcode1,bcode2)\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\")\n",
    "r3= barcode_compare(bcode1,bcode2,method=\"other\")\n",
    "\n",
    "\n",
    "print(\"One mismatch r1:\" + str(r1))\n",
    "print(\"One mismatch r2:\" + str(r2))\n",
    "print(\"One mismatch r3:\" + str(r3))\n",
    "\n",
    "\n",
    "\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGACG\"\n",
    "r1 = barcode_compare(bcode1,bcode2)\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\")\n",
    "\n",
    "print(\"One indel r1:\" + str(r1))\n",
    "print(\"One indel r2:\" + str(r2))\n",
    "\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGCANG\"\n",
    "r1 = barcode_compare(bcode1,bcode2)\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\")\n",
    "r3= barcode_compare(bcode1,bcode2,method=\"other\")\n",
    "\n",
    "\n",
    "print(\"One mismatch with 1N r1:\" + str(r1))\n",
    "print(\"One mismatch with 1N r2:\" + str(r2))\n",
    "print(\"One mismatch with 1N r3:\" + str(r3))\n",
    "\n",
    "\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGCATG\"\n",
    "r1 = barcode_compare(bcode1,bcode2)\n",
    "r2 = barcode_compare(bcode1,bcode2,method=\"difflib\")\n",
    "r3= barcode_compare(bcode1,bcode2,method=\"other\")\n",
    "print(\"Two mismatch r1:\" + str(r1))\n",
    "print(\"Two mismatch r2:\" + str(r2))\n",
    "print(\"Two mismatch r3:\" + str(r3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Speed will be very different given different comparison methods. Let's test the computation time. From the results we could tell the running time relationship is \n",
    " ** DP = 10 $\\times$ difflib $\\times$ others**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computation time for 10 + DP is 0.00704383850098s\n",
      "computation time for 10 + difflib is 0.00076699256897s\n",
      "computation time for 10 + other is 7.5101852417e-05s\n",
      "computation time for 100 + DP is 0.072448015213s\n",
      "computation time for 100 + difflib is 0.00754690170288s\n",
      "computation time for 100 + other is 0.000773906707764s\n",
      "computation time for 1000 + DP is 0.468641996384s\n",
      "computation time for 1000 + difflib is 0.0468771457672s\n",
      "computation time for 1000 + other is 0.00502896308899s\n",
      "computation time for 10000 + DP is 4.714812994s\n",
      "computation time for 10000 + difflib is 0.373748064041s\n",
      "computation time for 10000 + other is 0.032851934433s\n"
     ]
    }
   ],
   "source": [
    "# this is a test to see the speed of the barcode_compare.\n",
    "import time\n",
    "bcode1 = \"ACTGGACG\"\n",
    "bcode2 = \"ACTGNACG\"\n",
    "methods = [\"DP\",\"difflib\",\"other\"]\n",
    "compare_times = [10,100,1000,10000]\n",
    "for h in range(len(compare_times)):\n",
    "    compare_time = compare_times[h]\n",
    "    for i in range(len(methods)):\n",
    "        method = methods[i]\n",
    "        tic = time.time()\n",
    "        for j in range(compare_time):\n",
    "            r = barcode_compare(bcode1,bcode2,method=method)\n",
    "        tac = time.time()\n",
    "        delta = tac - tic\n",
    "        print (\"computation time for {} + {} is {}s\".format(compare_time,method,delta))\n",
    "# print (barcode_count)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our algorithm will cluster the barcode together if one barcode is similar to another by certain cut-off. The greedy algorithm is designed to cluster the barcode with lower count into barcode with higher count. There are some technical details should be noticed:\n",
    "* for each probe barcode is ordered accroding to the count both asc and descly.\n",
    "* from the lowest count barcode, barcode will compare to from the highest count barcode to decide whether to merge.\n",
    "* speed will vary in different methods. The speed of greed_cluster is $O(N^2)$\n",
    "* to reduce the computational burden, we introduce $desc\\_threshold$ and $asc\\_threshold$. (See comment for the meaning)\n",
    "* $score1\\_threshold$ and $score2\\_threshold$ should be modified according to different barcode_compare_methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to perform greedy cluster.\n",
    "def greedy_cluster (barcode_count, barcode_compare_method = \"other\", score1_threshold = 6, score2_threshold = 6, desc_threshold = 99999, asc_threshold = 99999):\n",
    "    '''\n",
    "    Input: \n",
    "    barcode_count is a dictionary. key is the barcode pair, and\n",
    "    value is the count of this pair of barcode on the probe.\n",
    "    \n",
    "    e.g. barcode_count = {('ACTGGACG', 'ACTGCGTT'): 100, ('ACTGGACG', 'ACTGGACGC'): 5,('ACNGGACG','TCATGACG'):2}\n",
    "    \n",
    "    barcode_compare_method is the method to calculate pair-wise barcode similarity.\n",
    "\n",
    "    score1_threshold is 5'end comparison lower bound.\n",
    "    score2_threshold is 3'end comparison lower bound.\n",
    "    \n",
    "    desc_threshold is the number of candidate barcodes to be merged into. - i.e. added.\n",
    "    asc_threshold is the candidate of candidate barcodes to be merged from - i.e. deleted.\n",
    "    \n",
    "    Output: \n",
    "    updated barcode_count.\n",
    "    '''\n",
    "    \n",
    "    from collections import OrderedDict\n",
    "\n",
    "    barcode_count_asc = OrderedDict(sorted(barcode_count.items(), key=lambda kv: kv[1]))\n",
    "    barcode_count_desc = OrderedDict(sorted(barcode_count.items(), key=lambda kv: -kv[1]))\n",
    "\n",
    "    # define the candidate query set need merged.\n",
    "    q_pair_set = barcode_count_asc.keys()\n",
    "    \n",
    "    for i in range(min(asc_threshold, len(q_pair_set))):\n",
    "                \n",
    "        q5 = q_pair_set[i][0]\n",
    "        q3 = q_pair_set[i][1]\n",
    "        count = barcode_count_asc[q_pair_set[i]]\n",
    "        \n",
    "        # do not compare with itself\n",
    "        del barcode_count_desc[q_pair_set[i]]\n",
    "        \n",
    "        # the reference dict is not empty.\n",
    "        if barcode_count_desc:\n",
    "            \n",
    "            # define the candidate reference set could be merged into.\n",
    "            r_pair_set = barcode_count_desc.keys()\n",
    "\n",
    "            for j in range(min(desc_threshold, len(r_pair_set))):\n",
    "                r5 = r_pair_set[j][0]\n",
    "                r3 = r_pair_set[j][1]\n",
    "                \n",
    "                score1 = barcode_compare(q5,r5,method=barcode_compare_method)\n",
    "                score2 = barcode_compare(q3,r3,method=barcode_compare_method)        \n",
    "                \n",
    "                if score1 >= score1_threshold and score2 >= score2_threshold:\n",
    "                    \n",
    "                    barcode_count[r_pair_set[j]] += count\n",
    "                    del barcode_count[q_pair_set[i]]\n",
    "                    \n",
    "                    break\n",
    "    \n",
    "    return (barcode_count)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computation time for 10 is 0.0177750587463s\n",
      "{('TAATGGCC', 'AAAGCTCG'): 2, ('GCCTCTAC', 'AGAGGTCA'): 4, ('GATAAGCT', 'CAACCGAC'): 2, ('ACCCGAGG', 'TACTACAC'): 2, ('GTACTGGA', 'GACACCAA'): 2, ('CTGCGTCC', 'GGACCCGA'): 4}\n"
     ]
    }
   ],
   "source": [
    "# this is a test to see the speed of the greedy_cluster.\n",
    "import time\n",
    "Ns = [10] # barcode_number\n",
    "for h in range(len(Ns)):\n",
    "    N = Ns[h]\n",
    "    tic = time.time()\n",
    "    for i in range(N*N/2): # O(N^2) \n",
    "        barcode_count_orig = {('GCCTCTAC', 'AGAGGGCA'): 2, ('TAATGGCC', 'AAAGCTCG'): 2, ('GCCTCTAC', 'AGAGGTCA'): 2, ('GATAAGCT', 'CAACCGAC'): 2, ('CTGCGTCC', 'GGACCCGA'): 4, ('GTACTGGA', 'GACACCAA'): 2, ('ACCCGAGG', 'TACTACAC'): 2}\n",
    "        barcode_count_new = greedy_cluster(barcode_count_orig)\n",
    "    tac = time.time()\n",
    "    delta = tac - tic\n",
    "    print (\"computation time for {} is {}s\".format(N,delta))\n",
    "print (barcode_count_new)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the test above we could see the computation could be huge - ** $O(N^2)$ **\n",
    "for N = 1000, it takes ~1min to compute.\n",
    "Since in each dataset, we are expected to have 100,000,000 reads for one sample, and thus expect to have ~100,000 reads for each probe. The computing time could be larger than 1 year. ( The true time will be much less than that due to reads are not evenly distributed across barcodes).\n",
    "\n",
    "** Threfore, in practce, we have to optimize it **.\n",
    "* add a cut-off desc_threshold: only query top ranked barcode_pair in barcode_count_desc dictionary. (implmented already)\n",
    "* add a cut-off asc_threshold: only query top ranked barcode_pair in barcode_count_asc dictionary. (implmented already)\n",
    "* optimize the compare algorithm. (implmented already)\n",
    "    * DP = 10  ××  difflib  ××  others"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we implement read fastq and read sam file. \n",
    "\n",
    "Below is an example for a ** SAM ** line. SAM is ordered as fastq reads order.\n",
    "\n",
    "** TPNB500121:117:H7CLYBGX2:1:11101:17864:1043     73      MLH1_3_mc       1       42      30M     =       1       0\n",
    "       ACGCCAAAATATCGTTCGTAACAAAAATTA  AEEAEEEEEEEEEEEEEEEEEEEEAEEEEA  AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:30 YT:Z:UP**\n",
    "\n",
    "The raw fastq seq for this reads is \n",
    "\n",
    "** TACTANTAACGCCAAAATATCGTTCGTAACAAAAATTA **\n",
    "\n",
    "For this reads the barcode is ** TACTANTA **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to read and parse fastq\n",
    "def read_fastq (fq1,fq2,barcode_length = 8):\n",
    "    '''\n",
    "    Input: \n",
    "    a pair of fastq file fq1 and fq2\n",
    "    length of barcode\n",
    "    Output:\n",
    "    a dictionary key is reads id and value is [barcode1,barcode2]\n",
    "    '''\n",
    "    import gzip\n",
    "    from Bio import SeqIO\n",
    "    barcode_dict = {}\n",
    "    for fq in [fq1,fq2]:\n",
    "        with gzip.open(fq) as f:\n",
    "            records = SeqIO.parse(f, \"fastq\")\n",
    "            for record in records:\n",
    "                seq_id = record.id\n",
    "                seq_barcode = str(record.seq)[:barcode_length]\n",
    "                if seq_id in barcode_dict.keys():\n",
    "                    barcode_dict[seq_id].append(seq_barcode)\n",
    "                else:\n",
    "                    barcode_dict[seq_id] = []\n",
    "                    barcode_dict[seq_id].append(seq_barcode)\n",
    "                    \n",
    "    return barcode_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TPNB500121:117:H7CLYBGX2:1:11101:21594:7682': ['ATTGGTTA', 'ATCGGTTA'], 'TPNB500121:117:H7CLYBGX2:1:11101:25651:7457': ['NNNNNNNN', 'ATCTCATA']}\n"
     ]
    }
   ],
   "source": [
    "# an example to show the function above.\n",
    "fq1 = \"/Users/cong/Desktop/barcode/SEQ4416_1_40000.fastq.gz\"\n",
    "fq2 = \"/Users/cong/Desktop/barcode/SEQ4416_2_40000.fastq.gz\"\n",
    "barcode_dict = read_fastq(fq1,fq2,8)\n",
    "first2pairs = {k: barcode_dict[k] for k in barcode_dict.keys()[:2]}\n",
    "print(first2pairs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** read_fastq ** function is to read and parse fastq data. The output is a dictionary where key is the reads id and value is a list of 5' and 3' barcode. Something have to be pay attention.\n",
    "* The RAM request is huge to store the whole fastq in memory.\n",
    "* Biopython should be installed.\n",
    "* What to do with NNNNNNNN?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to read and parse SAM\n",
    "def read_bam (bam_file,barcode_dict, output = 'everything'):\n",
    "    '''\n",
    "    Input:\n",
    "    bam_file. only bam file. SAM file does not work\n",
    "    barcode_dict obtained from read_fastq()\n",
    "    output_barcode: Should it output barcode or reads count.\n",
    "    barcode_details: Should it ouput a summary of barcode count for each probe or details.\n",
    "    Output:\n",
    "    bc_dict: a dictionary containning the count of barcode for each probe after greed_clustered.\n",
    "    '''\n",
    "    # change it from sam to bam.\n",
    "#     if filename.split(\".\")[-1] == \"sam\":\n",
    "#         pysam.sort(\"-o\", bam_file+\"sam\", bam_file)\n",
    "        \n",
    "    # creat an index for bam file for the purposes of random access \n",
    "    pysam.index(bam_file) \n",
    "    \n",
    "    f = pysam.AlignmentFile(bam_file,'rb')\n",
    "    \n",
    "    barcode_detail_dict = {}\n",
    "    barcode_detail_dict[\"pm\"] = {}\n",
    "    barcode_detail_dict[\"mm\"] = {}\n",
    "    barcode_detail_dict[\"mu\"] = {}\n",
    "    \n",
    "    uniq_barcode_dict = {}\n",
    "    uniq_barcode_dict[\"pm\"] = {}\n",
    "    uniq_barcode_dict[\"mm\"] = {}\n",
    "    uniq_barcode_dict[\"mu\"] = {}\n",
    "    \n",
    "    reads_dict = {}\n",
    "    reads_dict[\"pm\"] = {}\n",
    "    reads_dict[\"mm\"] = {}\n",
    "    reads_dict[\"mu\"] = {}\n",
    "    \n",
    "    for probe in (f.references):\n",
    "        barcode_count_pm = {} # both ends mapped to the same ref.\n",
    "        barcode_count_mm = {} # mate maps to different ref.\n",
    "        barcode_count_mu = {} # mate unmapped.\n",
    "        reads = f.fetch(probe)\n",
    "        sam = []\n",
    "        for read in reads:\n",
    "            barcode = barcode_dict[read.query_name]\n",
    "            barcode = tuple(barcode) # list is not hashaable.\n",
    "# the reads status could be the following.\n",
    "# see http://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.aligned_pairs for details.\n",
    "#             read.mate_is_reverse\n",
    "#             read.mate_is_unmapped\n",
    "#             read.is_unmapped\n",
    "#             read.is_duplicate\n",
    "#             read.is_paired\n",
    "#             read.is_proper_pair\n",
    "#             read.is_qcfail\n",
    "#             read.is_read1\n",
    "#             read.is_read2\n",
    "#             read.is_reverse\n",
    "#             read.is_secondary\n",
    "#            sam.append(str(read))\n",
    "\n",
    "            if read.is_proper_pair:\n",
    "                if barcode in barcode_count_pm.keys():\n",
    "                    barcode_count_pm[barcode] += 1\n",
    "                else:\n",
    "                    barcode_count_pm[barcode] = 1\n",
    "                \n",
    "            else:\n",
    "                if not (read.is_unmapped):\n",
    "                    if read.mate_is_unmapped:\n",
    "                        if barcode in barcode_count_mu.keys():\n",
    "                            barcode_count_mu[barcode] += 1\n",
    "                        else:\n",
    "                            barcode_count_mu[barcode] = 1\n",
    "                    else:\n",
    "                        if barcode in barcode_count_mm.keys():\n",
    "                            barcode_count_mm[barcode] += 1\n",
    "                        else:\n",
    "                            barcode_count_mm[barcode] = 1\n",
    "                else:\n",
    "                    next # don't count unmapped reads since it already counted by its mate.\n",
    "        # print len(barcode_count.keys())\n",
    "        barcode_detail_dict['pm'][probe] = greedy_cluster(barcode_count=barcode_count_pm,asc_threshold=50,desc_threshold=50)\n",
    "        barcode_detail_dict['mm'][probe] = greedy_cluster(barcode_count=barcode_count_mm,asc_threshold=50,desc_threshold=50)\n",
    "        barcode_detail_dict['mu'][probe] = greedy_cluster(barcode_count=barcode_count_mu,asc_threshold=50,desc_threshold=50)\n",
    "        \n",
    "        uniq_barcode_dict['pm'][probe] = len(barcode_detail_dict['pm'][probe].keys())\n",
    "        uniq_barcode_dict['mm'][probe] = len(barcode_detail_dict['mm'][probe].keys())\n",
    "        uniq_barcode_dict['mu'][probe] = len(barcode_detail_dict['mu'][probe].keys())\n",
    "        \n",
    "        reads_dict[\"pm\"][probe] = sum(barcode_detail_dict['pm'][probe].values())/2 # a pair will only count once. \n",
    "        reads_dict[\"mm\"][probe] = sum(barcode_detail_dict['mm'][probe].values())\n",
    "        reads_dict[\"mu\"][probe] = sum(barcode_detail_dict['mu'][probe].values())\n",
    "\n",
    "        \n",
    "#         reads_dict[probe] = sam\n",
    "    if output == 'barcode count':\n",
    "        return barcode_detail_dict\n",
    "    elif output == \"read count\":\n",
    "        return reads_dict\n",
    "    elif output == \"barcode detail\":\n",
    "        return uniq_barcode_dict\n",
    "    else:\n",
    "        return barcode_detail_dict,reads_dict,uniq_barcode_dict\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a program to test the read_bam.\n",
    "import time\n",
    "import pysam\n",
    "\n",
    "tic = time.time()\n",
    "sam_file = \"/Users/cong/Desktop/barcode/SEQ4416_10112.sam\"\n",
    "bam_file = \"/Users/cong/Desktop/barcode/SEQ4416_10112.bam\"\n",
    "# pysam.sort(\"-o\", bam_file, sam_file)\n",
    "barcode_detail_dict, reads_dict, uniq_barcode_dict = read_bam(bam_file,barcode_dict)\n",
    "\n",
    "# print reads_dict['MAL_1_mc']\n",
    "tac = time.time()\n",
    "# print (tac - tic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$read\\_bam$ is a function to read bam and count the barcode count for each probe. it could \n",
    "* return a dictionary containning count for each barcodes and each probe\n",
    "* OR return a dictionary containing total barcode count for each probe.\n",
    "* OR return a dictionary containing total reads count for each probe.\n",
    "\n",
    "Three types of reads count are returned.\n",
    "* pm: read is mapped in this probe in a proper pair (a pair will only count once)\n",
    "* mm: one end of read is mapped in this probe, another in another or discordordantly\n",
    "* mu: one end of read is mapped in this probe, another unmapped\n",
    "\n",
    "** pysam module is required to run the code.**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we implement main function to execute the barcode count program."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def main(bam_file,fq1,fq2,output):\n",
    "    \n",
    "    import time\n",
    "    import pysam\n",
    "    from collections import OrderedDict\n",
    "    import difflib as df\n",
    "    import numpy as np\n",
    "    from numpy import array\n",
    "    import re\n",
    "    import csv\n",
    "    import pandas as pd\n",
    "    \n",
    "    barcode_dict = read_fastq(fq1,fq2,8)\n",
    "    barcode_detail_dict, reads_dict, uniq_barcode_dict = read_bam(bam_file,barcode_dict)\n",
    "    \n",
    "#     count_df = pd.DataFrame(count)\n",
    "#     count_df.to_csv(output)\n",
    "#     with open(output, 'wb') as csv_file:\n",
    "#         csv_writer = csv.writer(csv_file)\n",
    "#         for key, value in count.items():\n",
    "#             csv_writer.writerow([key, value])\n",
    "        \n",
    "    return barcode_detail_dict, reads_dict, uniq_barcode_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TPNB500121:117:H7CLYBGX2:1:11101:12199:1424\t['GTACTGGA', 'GACACCAA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:6005:1606\t['CTGCGTCC', 'GGACCCGA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:25086:1858\t['GCCTCTAC', 'AGAGGTCA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:12066:2614\t['GTTAAGCG', 'GTTTAGCG']\tCDH13_2_mc\tFalse\tFalse\tTrue\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:12066:2614\t['GTTAAGCG', 'GTTTAGCG']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:20480:2932\t['TAATGGCC', 'AAAGCTCG']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:6782:3708\t['CTGCGTCC', 'GGACCCGA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:25630:4174\t['ACCCGAGG', 'TACTACAC']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:16199:4357\t['GATAAGCT', 'CAACCGAC']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:18196:4543\t['GCCTCTAC', 'AGAGGGCA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:11492:4570\t['GTGTTTGG', 'GTGTTTAG']\tCDH13_2_mc\tFalse\tFalse\tTrue\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:11492:4570\t['GTGTTTGG', 'GTGTTTAG']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:11306:1588\t['GAGACGTT', 'AATGACTC']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:15236:4264\t['GAGACGTT', 'AATGACTC']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:12199:1424\t['GTACTGGA', 'GACACCAA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:12861:1458\t['CCGAATAT', 'ATAGTGCC']\tCDH13_2_mc\tFalse\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:11306:1588\t['GAGACGTT', 'AATGACTC']\tCDH13_2_mc\tFalse\tFalse\tTrue\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:6005:1606\t['CTGCGTCC', 'GGACCCGA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:25086:1858\t['GCCTCTAC', 'AGAGGTCA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:6081:2324\t['GTCTGGAG', 'ATTCTGGT']\tCDH13_2_mc\tFalse\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:20480:2932\t['TAATGGCC', 'AAAGCTCG']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:3747:2948\t['GTAGCGTT', 'CCTACGAT']\tCDH13_2_mc\tFalse\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:6782:3708\t['CTGCGTCC', 'GGACCCGA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:25630:4174\t['ACCCGAGG', 'TACTACAC']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:15236:4264\t['GAGACGTT', 'AATGACTC']\tCDH13_2_mc\tFalse\tFalse\tTrue\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:16199:4357\t['GATAAGCT', 'CAACCGAC']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:18196:4543\t['GCCTCTAC', 'AGAGGGCA']\tCDH13_2_mc\tTrue\tFalse\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:1873:3252\t['AAAAAAAA', 'AAGAAAAA']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:4557:5216\t['AAAAAAAA', 'AATAAGAA']\tCDH13_2_mc\tFalse\tTrue\tFalse\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:1873:3252\t['AAAAAAAA', 'AAGAAAAA']\tCDH13_2_mc\tFalse\tFalse\tTrue\n",
      "TPNB500121:117:H7CLYBGX2:1:11101:4557:5216\t['AAAAAAAA', 'AATAAGAA']\tCDH13_2_mc\tFalse\tFalse\tTrue\n"
     ]
    }
   ],
   "source": [
    "# a test program to see whether it returns the correct answer.\n",
    "\n",
    "import pysam\n",
    "f = pysam.AlignmentFile(bam_file,'rb')\n",
    "reads = f.fetch('CDH13_2_mc')\n",
    "barcode_dict = read_fastq(fq1,fq2,8)\n",
    "for read in reads:\n",
    "    print ('{}\\t{}\\t{}\\t{}\\t{}\\t{}').format(read.query_name,barcode_dict[read.query_name],read.reference_name,read.is_proper_pair,read.is_unmapped,read.mate_is_unmapped)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "sam_file = \"/Users/cong/Desktop/barcode/SEQ4416_10112.sam\"\n",
    "bam_file = \"/Users/cong/Desktop/barcode/SEQ4416_10112.bam\"\n",
    "fq1 = \"/Users/cong/Desktop/barcode/SEQ4416_1_40000.fastq.gz\"\n",
    "fq2 = \"/Users/cong/Desktop/barcode/SEQ4416_2_40000.fastq.gz\"\n",
    "output = \"/Users/cong/Desktop/barcode/SEQ4416_40000_barcode_count.csv\"\n",
    "barcode_detail_dict, reads_dict, uniq_barcode_dict = main(bam_file=bam_file,fq1=fq1,fq2=fq2,output=output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode details.\n",
      "----------------\n",
      "proper pair: {('TAATGGCC', 'AAAGCTCG'): 2, ('GCCTCTAC', 'AGAGGTCA'): 4, ('GATAAGCT', 'CAACCGAC'): 2, ('CTGCGTCC', 'GGACCCGA'): 4, ('GTACTGGA', 'GACACCAA'): 2, ('ACCCGAGG', 'TACTACAC'): 2}\n",
      "unproper pair: {('GTAGCGTT', 'CCTACGAT'): 1, ('GTCTGGAG', 'ATTCTGGT'): 1, ('CCGAATAT', 'ATAGTGCC'): 1}\n",
      "single map: {('AAAAAAAA', 'AATAAGAA'): 1, ('AAAAAAAA', 'AAGAAAAA'): 1, ('GAGACGTT', 'AATGACTC'): 2, ('GTTAAGCG', 'GTTTAGCG'): 1, ('GTGTTTGG', 'GTGTTTAG'): 1}\n",
      "barcode count.\n",
      "----------------\n",
      "proper pair: 6\n",
      "unproper pair: 3\n",
      "single map: 5\n",
      "reads count.\n",
      "----------------\n",
      "proper pair: 8\n",
      "unproper pair: 3\n",
      "single map: 6\n"
     ]
    }
   ],
   "source": [
    "print (\"barcode details.\")\n",
    "print (\"----------------\")\n",
    "print (\"proper pair: {}\").format(barcode_detail_dict['pm']['CDH13_2_mc'])\n",
    "print (\"unproper pair: {}\").format(barcode_detail_dict['mm']['CDH13_2_mc'])\n",
    "print (\"single map: {}\").format(barcode_detail_dict['mu']['CDH13_2_mc'])\n",
    "print (\"barcode count.\")\n",
    "print (\"----------------\")\n",
    "print (\"proper pair: {}\").format(uniq_barcode_dict['pm']['CDH13_2_mc'])\n",
    "print (\"unproper pair: {}\").format(uniq_barcode_dict['mm']['CDH13_2_mc'])\n",
    "print (\"single map: {}\").format(uniq_barcode_dict['mu']['CDH13_2_mc'])\n",
    "print (\"reads count.\")\n",
    "print (\"----------------\")\n",
    "print (\"proper pair: {}\").format(reads_dict['pm']['CDH13_2_mc'])\n",
    "print (\"unproper pair: {}\").format(reads_dict['mm']['CDH13_2_mc'])\n",
    "print (\"single map: {}\").format(reads_dict['mu']['CDH13_2_mc'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the manual check we could find that\n",
    "* **TPNB500121:117:H7CLYBGX2:1:11101:25086:1858** and **TPNB500121:117:H7CLYBGX2:1:11101:18196:4543** only have one mismatch on barcode so that they were merged together same barcode\n",
    "The results are manually checked correct.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Issues:\n",
    "\n",
    "* manual check in small samples and compare with other tools (i.e. feature count) \n",
    "* test one sample in server.\n",
    "* discuss w/ groups for next steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pm: \n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:12199:1424\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:6005:1606\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:25086:1858\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:20480:2932\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:6782:3708\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:25630:4174\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:16199:4357\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:18196:4543\n",
    "\n",
    "\n",
    "\n",
    "mu:\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:12066:2614\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:11492:4570\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:11306:1588\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:15236:4264\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:1873:3252\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:4557:5216\n",
    "\n",
    "mm:\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:12861:1458 - 145\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:6081:2324\n",
    "* TPNB500121:117:H7CLYBGX2:1:11101:3747:2948\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
